### PR TITLE
fix issue 2485 which occur oom when using async servlet request.

### LINKED ISSUE #2485: OOM occured with StringResourceWrapper linked CtEntry in Tomcat TaskThread ThreadLocal.
## Issue Description

Type: *bug report* or *feature request*

### Describe what happened (or what feature you want)

guess something happened in afterCompletion, Context or Entry not clean and exit.
TaskThread in tomcat is used in a pool, new request get a old Context from pre request.
of course this is a guess，because i can't get the code for debug.

```
public void afterCompletion(HttpServletRequest request, HttpServletResponse response,
                                Object handler, Exception ex) throws Exception {
        if (increaseReferece(request, this.baseWebMvcConfig.getRequestRefName(), -1) != 0) {
            return;
        }
        
        Entry entry = getEntryInRequest(request, baseWebMvcConfig.getRequestAttributeName());
        if (entry == null) {
            // should not happen
            RecordLog.warn("[{}] No entry found in request, key: {}",
                    getClass().getSimpleName(), baseWebMvcConfig.getRequestAttributeName());
            return;
        }
        
        traceExceptionAndExit(entry, ex);
        removeEntryInRequest(request);
        ContextUtil.exit();
    }
```

image
image

image

### Describe what you expected to happen

ContextUtil map contains 2 key（default、spring-sentinel）is no problem.
It maybe occured in interceptor which has no close action.

### How to reproduce it (as minimally and precisely as possible)

1. 
2. 
3. 

### Tell us your environment

version 1.8.2
pom:spring-cloud-alibaba
pom:webmvc-adapter
no rule used and default config

### PR DESCRIPTION
**Pull request by Linzyoo**
### Describe what this PR does / why we need it
There is an issue about it.

### Does this pull request fix one issue?

Fixes #2485

### Describe how you did it
Extends HandlerInterceptor with a callback method invoked after the start of asynchronous request handling.
When a handler starts an asynchronous request, the DispatcherServlet exits without invoking postHandle and afterCompletion. Then it will not exit the context and clean the thread-local variables for the request.  So I extends AsynHandlerInterceptor which is to deal with this situation correctly.

### Describe how to verify it
Debug to check context and ctentry.
Add a unit test for Async servlet request, and check the context after response.

### CODE REVIEW

File: sentinel-demo/sentinel-demo-spring-webmvc/src/main/java/com/alibaba/csp/sentinel/demo/spring/webmvc/controller/WebMvcTestController.java [hunk 1]
```diff
@@ -65,6 +67,17 @@ public ModelAndView apiForward() {
         return mav;
     }
 
+    @GetMapping("/async")
+    @ResponseBody
+    public DeferredResult<String> distribute() throws Exception{
+//        return new DeferredResult<String>(5000L, (Supplier<String>) () -> {doBusiness(); return "err";});
```
Comment (LearningGp) [inline sentinel-demo/sentinel-demo-spring-webmvc/src/main/java/com/alibaba/csp/sentinel/demo/spring/webmvc/controller/WebMvcTestController.java:73]: Removing unused code would be better.

File: sentinel-adapter/sentinel-spring-webmvc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/spring/webmvc/AbstractSentinelInterceptor.java [hunk 2]
```diff
@@ -135,21 +137,43 @@ protected String getContextName(HttpServletRequest request) {
...
     public void afterCompletion(HttpServletRequest request, HttpServletResponse response,
                                 Object handler, Exception ex) throws Exception {
+        exitContext(request, ex);
+    }
+
+    private void exitContext(HttpServletRequest request) {
+        exitContext(request, null);
+    }
+
+    private void exitContext(HttpServletRequest request, Exception ex) {
```
Comment (LearningGp) [inline sentinel-adapter/sentinel-spring-webmvc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/spring/webmvc/AbstractSentinelInterceptor.java:164]: This function does more than just exiting the context, so naming it simply 'exit' might be more appropriate.

**Review by LearningGp — CHANGES_REQUESTED** [review]

File: sentinel-adapter/sentinel-spring-webmvc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/spring/webmvc/AbstractSentinelInterceptor.java (hunk 2 above)
Comment (Linzyoo) [reply sentinel-adapter/sentinel-spring-webmvc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/spring/webmvc/AbstractSentinelInterceptor.java:164]: ok

File: sentinel-demo/sentinel-demo-spring-webmvc/src/main/java/com/alibaba/csp/sentinel/demo/spring/webmvc/controller/WebMvcTestController.java (hunk 1 above)
Comment (Linzyoo) [reply sentinel-demo/sentinel-demo-spring-webmvc/src/main/java/com/alibaba/csp/sentinel/demo/spring/webmvc/controller/WebMvcTestController.java:73]: ok

**Comment by Linzyoo** [comment]
finished imporving codes based on the review comments. And update the webmvc-v6x too which has the same issue.

**Review by LearningGp — APPROVED** [review]
LGTM

**Pull request merged**